vcpkg/src/lib.rs

```diff
@@ -398,19 +398,196 @@ fn find_vcpkg_target(cfg: &Config, target_triplet: &TargetTriplet) -> Result<Vcp
+            }));
+            // Only look at .pc files.
+            if dir_entry.path().extension() != Some(OsStr::new("pc")) {
+                continue;
+            }
+            let pc_file = try!(PcFile::parse_pc_file(vcpkg_target, &dir_entry.path()));
+            files.insert(pc_file.id.to_owned(), pc_file);
+        }
+        Ok(PcFiles { files: files })
+    }
+    /// Use the .pc files as a hint to the library sort order.
+    fn fix_ordering(&self, mut libs: Vec<String>) -> Vec<String> {
+        // Overall heuristic: for each library given as input, identify which PcFile declared it.
+        // Then, looking at that PcFile, check its Requires: (deps), and if the pc file for that
+        // dep is in our set, check if its libraries are in our set of libs.  If so, move it to the
+        // end to ensure it gets linked afterwards.
+
+        // We may need to do this a few times to properly handle the case where A -> (depends on) B
+        // -> C -> D and libraries were originally sorted D, C, B, A.  Avoid recursion so we don't
+        // have to detect potential cycles.
+        let mut tries_left = 3;
+        loop {
+            let mut required_lib_order: Vec<String> = Vec::new();
+            for lib in &libs {
+                required_lib_order.push(lib.to_owned());
+                if let Some(pc_file) = self.locate_pc_file_by_lib(lib) {
+                    // Consider its requirements:
+                    for dep in &pc_file.deps {
+                        // Only consider pkgconfig dependencies we know about.
+                        let dep_pc_file = match self.files.get(dep) {
+                            None => continue,
+                            Some(pc_file) => pc_file,
+                        };
+                        // Intra-port library ordering found, pivot any already seen dep_lib to the
+                        // end of the list.
+                        for dep_lib in &dep_pc_file.libs {
+                            if let Some(removed) = remove_item(&mut required_lib_order, dep_lib) {
+                                required_lib_order.push(removed);
+                            }
+                        }
+                    }
+                }
+            }
+            // We should always end up with the same number of libraries, only their order should
+            // change.
+            assert_eq!(libs.len(), required_lib_order.len());
+            // Termination:
+            if required_lib_order == libs {
+                // Nothing changed, we're done here.
+                return libs;
+            } else {
+                // Only try a few times.
+                if tries_left <= 0 {
+                    println!("cargo:warning=vcpkg gave up trying to resolve pkg-config ordering.");
+                    return libs;
+                } else {
+                    libs = required_lib_order;
+                    tries_left = tries_left - 1;
+                    continue;
```

Comment:
This `continue` is superfluous, I think? It might also simplify things to have the loop abort by just making it a `for _iter in 0..3 { ... }`.

Reply:
Yes that made this a lot easier to follow, thanks!